Add block examples for all core blocks

youknowriad opened the pull request:
closes #17488

This PR adds block examples for the core blocks in order to show a preview of the blocks in the inserter help panel. It uses the examples in the #17488 issue.

<img width="782" alt="Screenshot 2019-09-23 at 13 07 58" src="https://user-images.githubusercontent.com/548849/65421987-7b849500-de05-11e9-8593-a93631bd94ff.png">

![image](https://user-images.githubusercontent.com/548849/65442215-0e382a80-de2c-11e9-82b2-5f5e958cc7a3.png)

youknowriad committed:
Add block examples for all core blocks

Soean commented on packages/block-library/src/rss/index.js:
```diff
@@ -20,5 +20,10 @@ export const settings = {
 	},
+	example: {
+		attributes: {
+			feedURL: 'https://wordpress.org',
```
```suggestion
			feedURL: __( 'https://wordpress.org' ),
```

richtabor replied:
> Suggested change 
> 			feedURL: 'https://wordpress.org',
> 			feedURL: __( 'https://wordpress.org' ),

URL's don't need to be translated. 👌

HardeepAsrani replied:
It's usually helpful to translate users to redirect users to the homepage of their locale website like https://hi.wordpress.org/ if they're using WP in that locale.

Soean reviewed (commented):
I added some suggestions.

gziolo commented on packages/block-library/src/image/index.js:
```diff
@@ -25,6 +25,13 @@ export const settings = {
+		attributes: {
+			sizeSlug: 'large',
+			url: 'https://images.unsplash.com/photo-1549880339-d93e3072aef4',
+			caption: __( 'Snow covered mountain' ),
```
Image(s) from Unsplash probably need a copyright note.

mtias replied:
We should bundle any image we use as well.

youknowriad replied:
Can't we use images hosted on wp.org instead? Bundling could be challenging as the public URL is not easily accessible here.

gziolo commented on packages/blocks/src/api/factory.js:
```diff
@@ -461,3 +462,17 @@ export function switchToBlockType( blocks, name ) {
+ *
+ * @return {Object} block.
+ */
+export const getBlockFromExample = ( name, example ) => {
```
Should it be `createBlockFromExample` for conistency with the default name and the description?
```suggestion
export const createBlockFromExample = ( name, example ) => {
```

mtias committed:
Update "Quote" example.

mtias committed:
Fix "Group" example.

mtias committed:
Use a cover image for "Cover"; update heading text, and more fixes.

mtias committed:
Use Don Quijote excerpt for paragraph block with drop-cap.

mtias committed:
Cover image could use a bit more personality.

mtias committed:
Use "marquee" for the custom HTML block.

mtias committed:
Make "download WordPress" text translatable.

Co-Authored-By: Sören Wrede <soerenwrede@gmail.com>

mtias commented:
This is a great place to add some personality.

mtias committed:
Shelley.

mtias committed:
Elizabeth Barrett Browning.

mcsf commented on packages/block-library/src/video/index.js:
```diff
@@ -21,6 +21,12 @@ export const settings = {
 	keywords: [ __( 'movie' ) ],
+	example: {
+		attributes: {
+			src: 'https://make.wordpress.org/design/files/2019/03/block-manager.mp4',
```
This doesn't play in Safari.

mtias replied:
Is there any way to have a very small auto-playing video instead? It doesn't make sense to load a video that you cannot play.

jasmussen replied:
So, I found this place for public domain video: https://www.pond5.com/stock-video-footage/1/.html#1/18447/durationlt:10,pricelt:0

This one is cute, albeit short: https://www.pond5.com/stock-footage/44576423/little-smokey-walking-cage.html

But the problem then becomes, can we hotlink it? Where should these things be hosted? Is make.wordpress.org okay?

mtias committed:
Add instructions for creating a block in the "Code" block.

mtias committed:
Add Citizen Kane opening script lines for "preformatted".

mtias committed:
Use different colorful preview for "Group" block.

mtias committed:
Show "columns" in a better light.

mcsf commented on packages/block-library/src/columns/index.js:
```diff
@@ -24,8 +24,51 @@ export const settings = {
+				],
+			},
+		],
+	},
```
Needs overflow taken care of.

<img width="366" alt="Captura de ecrã 2019-09-23, às 14 19 05" src="https://user-images.githubusercontent.com/150562/65428915-29477200-de0d-11e9-984a-3bc97c8a7033.png">

mtias committed:
Add example for "Gallery" block.

mtias committed:
Update button example.

mtias committed:
Alice in Wonderland characters.

mcsf committed:
List: Example: Condense & optimise for rhythm

jasmussen commented:
This is really nice. Thank you for doing this. 

This marquee: 
![Screenshot 2019-09-23 at 16 12 58](https://user-images.githubusercontent.com/1204802/65433447-4e43e100-de1d-11e9-90da-4be8458708d3.png)

:slow clap:

Not sure if it's something for this PR, probably not, but there's a horizontal scrollbar here:

![Screenshot 2019-09-23 at 16 12 38](https://user-images.githubusercontent.com/1204802/65433477-5ef45700-de1d-11e9-950d-c91405a6e35d.png)

And this is possibly a theme issue (CC: @kjellr for insights):

![Screenshot 2019-09-23 at 16 12 24](https://user-images.githubusercontent.com/1204802/65433492-6ddb0980-de1d-11e9-9382-de97bebee258.png)

it's a little low.

Many of those seem to be more specific to the preview component, or the theme styling them. So not a blocker for this PR probably. And this PR is 👌

mcsf committed:
Media & Text: Example: Wren -- a haiku.

mtias committed:
Use strong markup for cover image text.

mapk commented:
Can the font-sizes get closer to match the Paragraph block preview? Several of the block previews are very small.

* List block
* Preformatted block
* Verse block
* HTML block
* Code block

![panel-sizes](https://user-images.githubusercontent.com/617986/65444397-d063e380-dde4-11e9-85fb-865111146470.gif)

mcsf committed:
Heading: Example: Keep H-level below 1

youknowriad commented:
@mapk Preview works by adapting the size to the container. So if the block is wider, it can be shown smaller than the others. It's going to be hard to align the size of all the blocks. We can probably try to tweak a bit.

Are we satisfied with the examples so far? I'd like us to have at least something in, even if not perfect for WordPress 5.3.

Also, should we replace Unsplash photos with other photos without licensing issues and upload them to wp.org?

mcsf commented:
In reply to https://github.com/WordPress/gutenberg/pull/17493#issuecomment-534119100

> it's a little low.

I think this is due to the loading of the images and not knowing their size beforehand. @jasmussen, if you hover once, move the cursor away, and hover again, does the positioning change?

jasmussen commented:
@mcsf 

> I think this is due to the loading of the images and not knowing their size beforehand. @jasmussen, if you hover once, move the cursor away, and hover again, does the positioning change?

Short answer, yes. Now I think it's a race condition. I did this:

- Check out the branch
- Open the library
- Mouse over the block
- The content was vertically centered, but as the image "trickled in" it was vertically offset
- Mousing over a different block then back reset it because now the dimensions can be red and accurately positioned.

Perhaps just adding a "height" attribute shoudl fix this.

Because it was a thing about loading, though I wanted to record a GIF of it, so I cleared my cache and started recording a GIF, and now the image won't load at all. Not sure what I did wrong:

![preview](https://user-images.githubusercontent.com/1204802/65507441-3c6e4680-dece-11e9-9ef2-075c3c8bbf86.gif)

mapk commented:
As for the examples being used...

### Group block 
The preview doesn't make much sense to me.

![Screen Shot 2019-09-24 at 10 49 11 AM](https://user-images.githubusercontent.com/617986/65536850-096d8800-deb9-11e9-9bd4-8f12bba95e82.png)

### Columns block
Everything else is a bit more creative than simple Lorem Ipsum. Let's replace it with this Robert Frost poem instead.

```
<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Two roads diverged in a yellow wood, And sorry I could not travel both, And be one traveler, long I stood, And looked down one as far as I could, To where it bent in the undergrowth;</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Then took the other, as just as fair, And having perhaps the better claim, Because it was grassy and wanted wear; Though as for that the passing there, Had worn them really about the same,</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>And both that morning equally lay. In leaves no step had trodden black. Oh, I kept the first for another day! </p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Yet knowing how way leads on to way, I doubted if I should ever come back.</p>
<!-- /wp:paragraph -->

<!-- wp:image {"sizeSlug":"large"} -->
<figure class="wp-block-image size-large"><img src="https://images.unsplash.com/photo-1549880339-d93e3072aef4" alt=""/></figure>
<!-- /wp:image -->

<!-- wp:paragraph -->
<p>I shall be telling this with a sigh, Somewhere ages and ages hence: Two roads diverged in a wood, and I––I took the one less traveled by,<br>And that has made all the difference.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>– Robert Frost</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
```

mtias commented:
@mapk it seems you are missing colors:

![image](https://user-images.githubusercontent.com/548849/65543294-2b92f480-df11-11e9-9ee7-112b5f88086a.png)

mapk commented:
I wonder if we should instead do something like this for the Group block?

![Screen Shot 2019-09-24 at 12 51 40 PM](https://user-images.githubusercontent.com/617986/65545374-1777d480-deca-11e9-9576-96c8952c5653.png)


```
<!-- wp:group {"backgroundColor":"very-dark-gray"} -->
<div class="wp-block-group has-very-dark-gray-background-color has-background"><div class="wp-block-group__inner-container"><!-- wp:paragraph {"customTextColor":"#ffffff","customBackgroundColor":"#9258d8"} -->
<p style="background-color:#9258d8;color:#ffffff" class="has-text-color has-background">Block one.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"customTextColor":"#ffffff","customBackgroundColor":"#3e8fdc"} -->
<p style="background-color:#3e8fdc;color:#ffffff" class="has-text-color has-background">Block two.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"customTextColor":"#ffffff","customBackgroundColor":"#5bcc88"} -->
<p style="background-color:#5bcc88;color:#ffffff" class="has-text-color has-background">Block three.</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:group -->
```

mapk committed:
Base64 encoded images that I pulled from Pexels. Also revised the copy in the Columns block.

mtias commented:
I tried something very close to that but looked too overwhelming / forced in comparison, so settled with the colored text for now.

mapk commented:
> I tried something very close to that but looked too overwhelming / forced in comparison, so settled with the colored text for now.

I think you're right with a dark background. Maybe with a lighter background?

![Screen Shot 2019-09-26 at 2 53 18 PM](https://user-images.githubusercontent.com/617986/65727738-79247400-e06d-11e9-9b5b-e0a91731a6d2.png)

Also, I'm not sure why, but I don't get the colored text in my preview. And it's not registering with me visually as a "group".

mapk commented:
Or maybe using the text approach, but with a colored background on the Group block...

![Screen Shot 2019-09-26 at 2 58 28 PM](https://user-images.githubusercontent.com/617986/65727948-2d25ff00-e06e-11e9-8fa4-5a8b0a5fd290.png)

jasmussen commented:
I did not see colored text in twenty nineteen. Perhaps something with them registering custom color palettes.

mapk commented:
I'm using the Gutenberg Starter Theme, so not sure if that effects it either.

youknowriad commented:
I think we should not rely on color palettes for the example and only use custom colors because these work across themes.

youknowriad committed:
Revert "Base64 encoded images that I pulled from Pexels. Also revised the copy in the Columns block."

This reverts commit 6472d1c38ae5d304bf96b57ecc7901cf0b042426.

youknowriad commented:
I reverted the base64 commit because in my testing it increases the bundle size of the block-library script by 1Mb. Let's use URLs for now until we figure out a better approach to bundle these images in Core instead.

youknowriad committed:
Use custom colors instead of the color palette

youknowriad commented:
updated the group example to avoid the use of the color palette. It should work well across themes now.

jasmussen reviewed (commented):
I really like this. I think we should scramble to get it in. 

On very slow connections it does surface a few improvements we can make to the preview component., but since that's only the first time you hover a block and then never after, I would not call that a blocker. Furthermore, it drastically improves the usability for sighted users of the block library, and it helps encourage 3rd party block developers to include a preview, that it's an important feature to ship.

However I think we should make a quick change. We are hotlinking the images, and we are using Unsplash. The latters license while very liberal, the suggestion to credit photographers makes it perhaps 0.05% less liberal than a strict interpretation of the GPL, which suggests complete and total freedom.

So just in order to be totally kosher, I would recommend replacing all images with certified public domain images from Wikimedia. Using this list https://commons.wikimedia.org/wiki/Featured_Pictures_in_the_Public_Domain I would suggest we could use the following:

- https://commons.wikimedia.org/wiki/File:Windbuchencom.jpg
- https://commons.wikimedia.org/wiki/File:MarsSunset.jpg
- https://commons.wikimedia.org/wiki/File:MarsSunset.jpg
- https://commons.wikimedia.org/wiki/File:M104_ngc4594_sombrero_galaxy_hi-res.jpg
- https://commons.wikimedia.org/wiki/File:Glacial_lakes,_Bhutan.jpg
- https://commons.wikimedia.org/wiki/File:Mount_Hood_reflected_in_Mirror_Lake,_Oregon.jpg
- https://commons.wikimedia.org/wiki/File:Kuznetsk_Alatau_1.jpg
- https://commons.wikimedia.org/wiki/File:Sediment_off_the_Yucatan_Peninsula.jpg
- https://commons.wikimedia.org/wiki/File:Rocks_El_Torcal_de_Antequera_karst_Andalusia_Spain.jpg
- https://commons.wikimedia.org/wiki/File:Icy_shores_of_Lake_Michigan.jpg
- https://commons.wikimedia.org/wiki/File:Drymona_water_fall_and_pool_north_Euboea_Greece.jpg
- https://commons.wikimedia.org/wiki/File:Beech_and_ferns_in_Gullmarsskogen.jpg

An additional benefit is that Wikemedia explicitly permits hotlinking: https://commons.wikimedia.org/wiki/Commons:Reusing_content_outside_Wikimedia/technical

I hope to take a stab at replacing the URLs and returning. 

The overall content, other than the images, are great, though, works well:

![examples](https://user-images.githubusercontent.com/1204802/65879926-30342e80-e391-11e9-9732-82d806593e61.gif)

jasmussen committed:
Update images to be public domain from Wikimedia.

jasmussen commented:
Okay, pushed good fastloading guaranteed publicdomain images that look good. These five, for reference:

- https://commons.wikimedia.org/wiki/File:Biologia_Centrali-Americana_-_Cantorchilus_semibadius_1902.jpg
- https://commons.wikimedia.org/wiki/File:Windbuchencom.jpg
- https://commons.wikimedia.org/wiki/File:Glacial_lakes,_Bhutan.jpg
- https://commons.wikimedia.org/wiki/File:Sediment_off_the_Yucatan_Peninsula.jpg
- https://commons.wikimedia.org/wiki/File:MtBlanc1.JPG

Had to look for Mont Blanc and a specific bird due to the prose around them.

A couple more things before we can merge:

- https://github.com/WordPress/gutenberg/pull/17493/files — commit suggestion?
- https://github.com/WordPress/gutenberg/pull/17493/files#r327064974 — I'm looking for a public domain video that plays in Safari, if you have one let me know

I will hopefully be back.

jasmussen committed:
Update the video.

This video is smaller, waaay faster loading, and it's more generic than showing the Block Manager. Good for the preview, where you can't play back the video anyway.

jasmussen commented:
I just updated the video to be a very short and generic WordPress logo file. This is more appropriate for the preview where you can't even play it back. Plus the preview looks good:

![Screenshot 2019-09-30 at 15 55 37](https://user-images.githubusercontent.com/1204802/65885530-21527980-e39b-11e9-9006-495364ce4bac.png)

With the help of @kjellr I'm investigating why it doesn't play back in Safari, as noted in the reviews. It may be the Make server.

jasmussen committed:
Remove the video for now.

I have created #17655 to add it back.

jasmussen reviewed (approved):
As noted previously, I really like this. I think we should get it in.

I replaced all the images with Wikimedia images that are public domain. They also allow us to hotlink, so that helps. In addition, the files are small, so they load really fast compared the the previous images, and given they will be shown in scaled down previews the fidelity is fine. 

I also removed the video, because the one we had did not play in Safari, and it should be hosted on a CDN. I created #17655 to track this. 

What remains here is lovely, and I think it's good to go.

However I also did a bit of work on this, so would appreciate a sanity check, for example from @kjellr.

mapk commented:
Just created a Trac ticket to host images on w.org's CDN: https://core.trac.wordpress.org/ticket/48292

hacknug commented:
I know it's a minor issue but, could we address the sizing of `.block-editor-block-card` items changing from block to block? It's somewhat annoying trying to preview blocks and read their names and descriptions without having those in a fixed position.

Guessing it's due to flex-grow/shrink since the explicit sizing seems to be in place already.
